pkg/inframetadata/reporter.go

```diff
@@ -0,0 +1,149 @@
+	opts := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+		return zapcore.NewSamplerWithOptions(core, 10*time.Second, 1, 100)
+	})
+	return logger.WithOptions(opts)
+}
+
+// NewReporter creates a new host metadata reporter.
+// The reporter consumes pcommon.Resources through its 'Consume' method and merges them into payload.HostMetadata payloads.
+// It then periodically exporters these functions through the pusher with a specified period.
+func NewReporter(logger *zap.Logger, pusher Pusher, period time.Duration) (*Reporter, error) {
+	hostMap, err := hostmap.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build host map: %w", err)
+	}
+
+	return &Reporter{
+		logger:  createSampledLogger(logger),
+		hostMap: hostMap,
+		pusher:  pusher,
+		closeCh: make(chan struct{}),
+		ticker:  time.NewTicker(period),
+	}, nil
+}
+
+// checkResource to see if it should be used by default.
+// A resource is usable if 'AttributeDatadogHostUseAsMetadata' is true or shouldUseByDefault is true.
+func checkResource(res pcommon.Resource) (bool, error) {
+	shouldUse := shouldUseByDefault
+	if val, ok := res.Attributes().Get(AttributeDatadogHostUseAsMetadata); ok {
+		if val.Type() != pcommon.ValueTypeBool {
+			return false, fmt.Errorf("%q has type %q, expected \"Bool\"", AttributeDatadogHostUseAsMetadata, val.Type())
+		}
+		shouldUse = val.Bool()
+	}
+	return shouldUse, nil
+}
+
+// Consume a resource for host metadata reporting purposes.
+// The resource will be used only if it is usable (see 'checkResource') and it has a host attribute.
+func (r *Reporter) Consume(res pcommon.Resource) error {
+	if shouldUse, err := checkResource(res); err != nil {
+		return fmt.Errorf("failed to check resource: %w", err)
+	} else if !shouldUse {
+		// The resource should not be used for host metadata.
+		return nil
+	}
+
+	src, ok := attributes.SourceFromAttrs(res.Attributes())
+	if !ok {
+		r.logger.Warn("resource does not have host-identifying attributes", zap.Any("attributes", res.Attributes()))
+		return nil
+	}
+	if src.Kind != source.HostnameKind {
+		// The resource does not identify a host (e.g. serverless resource)
+		return nil
+	}
+
+	r.hostMap.Update(src.Identifier, res)
+	return nil
+}
+
+// Run the reporter to periodically exporte
+func (r *Reporter) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	for {
+		select {
+		case <-r.ticker.C:
+			// extract payloads from hostMap and report them.
+			for host, payload := range r.hostMap.Extract() {
+				r.logger.Info("Sending host metadata",
+					zap.String("host", host))
+				if err := r.pusher.Push(ctx, payload); err != nil {
+					r.logger.Error("Failed to send host metadata",
+						zap.String("host", host),
+						zap.Error(err),
+						zap.Any("payload", payload),
+					)
+				}
+			}
+		case <-r.closeCh:
+			cancel()
+			r.logger.Info("Stopped reporter")
+			return nil
+		}
+	}
+}
+
+// Stop the reporter.
+func (r *Reporter) Stop(_ context.Context) {
```

Comment:
Why do we need the argument?

Reply:
We don't need it now, I am just trying to be consistent with the usual component interface within the OpenTelemetry Collector. But we can leave it out for now if you think it is unnecessary and change the interface in the future.

Reply:
If it's not needed for anything, I don't think we should have it.